File: playwright_py/pages/InventoryPage.py
from typing import Final, List
from playwright.sync_api import Page, expect
from pages.base_page import BasePage
from utils import urls

class InventoryPage(BasePage):

    def __init__(self, page: Page):
        super().__init__(page)
        self.page_url_fragment = urls.inventory
        self.page = page

        self.backpack = page.locator('[data-test="add-to-cart-sauce-labs-backpack"]')
        self.bike_light = page.locator('[data-test="add-to-cart-sauce-labs-bike-light"]')
        self.bolt_t_shirt = page.locator('[data-test="add-to-cart-sauce-labs-bolt-t-shirt"]')
        self.fleece_jacket = page.locator('[data-test="add-to-cart-sauce-labs-fleece-jacket"]')
        self.lab_sonesie = page.locator('[data-test="add-to-cart-sauce-labs-onesie"]')
        self.shirt_red = page.locator('[data-test="add-to-cart-test.allthethings()-t-shirt-(red)"]')

        self.inventory_items: Final[List] = [
            self.backpack,
            self.bike_light,
            self.bolt_t_shirt,
            self.fleece_jacket,
            self.lab_sonesie,
            self.shirt_red,
        ]

        self.inventory_list = page.locator('[data-test="inventory-list"]')
        self.page_title = page.locator('[data-test="title"]')

        self.shopping_cart_link = page.locator('[data-test="shopping-cart-link"]')
        self.cart_badge = page.locator('[data-test="shopping-cart-badge"]')
        self.checkout_btn = page.locator('[data-test="checkout"]')
        self.continue_btn = page.locator('[data-test="continue"]')
        self.finish_btn = page.locator('[data-test="finish"]')
        self.back_to_products_btn = page.locator('[data-test="back-to-products"]')
        self.complete_header = page.locator('[data-test="checkout_complete_container"] h2')
        self.inventory_item_name = page.locator('[data-test="inventory-item-name"]')
        self.inventory_item_price = page.locator('[data-test="inventory-item-price"]')

        self.menu_button = page.locator('#react-burger-menu-btn')
        self.logout_link = page.locator('[data-test="logout-sidebar-link"]')
        self.burger_menu = page.locator('[data-test="open-menu"]')

        self.sort_dropdown = page.locator('select[data-test="product_sort_container"]')
        self.sort_low_to_high = 'lohi'
        self.sort_high_to_low = 'hilo'
        self.sort_name_az = 'az'
        self.sort_name_za = 'za'

    def check_is_on_inventory_page(self):
        expect(self.page).to_have_url(lambda url: self.page_url_fragment in url)
        expect(self.inventory_list).to_be_visible()

    async def add_all_visible_inventory_items(self):
        for item in self.inventory_items:
            if await item.is_visible():
                await item.click()

File: playwright_ts/pages/InventoryPage.ts
import { env, url } from 'sdk_automation';
import { expect, Page } from '@playwright/test';
import { BasePage } from '../pages';

export class InventoryPage extends BasePage {
  static readonly selectors = {
    addToCartBackpack: '[data-test="add-to-cart-sauce-labs-backpack"]',
    addToCartBikeLight: '[data-test="add-to-cart-sauce-labs-bike-light"]',
    addToCartBoltTShirt: '[data-test="add-to-cart-sauce-labs-bolt-t-shirt"]',
    addToCartFleeceJacket: '[data-test="add-to-cart-sauce-labs-fleece-jacket"]',
    addToCartOnesie: '[data-test="add-to-cart-sauce-labs-onesie"]',
    addToCartRedShirt: '[data-test="add-to-cart-test.allthethings()-t-shirt-(red)"]',

    removeBackpack: '[data-test="remove-sauce-labs-backpack"]',
    removeBikeLight: '[data-test="remove-sauce-labs-bike-light"]',
    removeBoltTShirt: '[data-test="remove-sauce-labs-bolt-t-shirt"]',
    removeFleeceJacket: '[data-test="remove-sauce-labs-fleece-jacket"]',
    removeOnesie: '[data-test="remove-sauce-labs-onesie"]',
    removeRedShirt: '[data-test="remove-test.allthethings()-t-shirt-(red)"]',

    inventoryList: '[data-test="inventory-list"]',
    pageTitle: '[data-test="title"]',
    shoppingCartLink: '[data-test="shopping-cart-link"]',
    cartBadge: '[data-test="shopping-cart-badge"]',
    inventoryItemName: '[data-test="inventory-item-name"]',
    inventoryItemPrice: '[data-test="inventory-item-price"]',
    sortDropdown: 'select[data-test="product-sort-container"]',
    logoutLink: '[data-test="logout-sidebar-link"]',
    burgerMenu: '[data-test="open-menu"]',
  };

  readonly page: Page;
  readonly pageUrl;
  readonly pageTitle;
  readonly inventoryList;
  readonly shoppingCartLink;
  readonly cartBadge;
  readonly inventoryItemName;
  readonly inventoryItemPrice;
  readonly sortDropdown;
  readonly inventoryItems;

  constructor(page: Page) {
    super(page);
    this.page = page;
    this.pageUrl = url.inventory;
    this.pageTitle = page.locator(InventoryPage.selectors.pageTitle);
    this.inventoryList = page.locator(InventoryPage.selectors.inventoryList);
    this.shoppingCartLink = page.locator(InventoryPage.selectors.shoppingCartLink);
    this.cartBadge = page.locator(InventoryPage.selectors.cartBadge);
    this.inventoryItemName = page.locator(InventoryPage.selectors.inventoryItemName);
    this.inventoryItemPrice = page.locator(InventoryPage.selectors.inventoryItemPrice);
    this.sortDropdown = page.locator(InventoryPage.selectors.sortDropdown);

    this.inventoryItems = [
      InventoryPage.selectors.addToCartBackpack,
      InventoryPage.selectors.addToCartBikeLight,
      InventoryPage.selectors.addToCartBoltTShirt,
      InventoryPage.selectors.addToCartFleeceJacket,
      InventoryPage.selectors.addToCartOnesie,
      InventoryPage.selectors.addToCartRedShirt,
    ];
  }

  // ==========================
  // Page Assertions / Checks
  // ==========================
  async checkIsOnInventoryPage(): Promise<void> {
    await expect.soft(this.page).toHaveURL(new RegExp(url.inventory));
    await expect.soft(this.pageTitle).toHaveText('Products');
    await expect.soft(this.sortDropdown).toBeVisible();
    await expect.soft(this.shoppingCartLink).toBeVisible();
    await expect.soft(this.inventoryList).toBeVisible();
  }

  // ==========================
  // Actions
  // ==========================
  async addAllVisibleInventoryItems(): Promise<void> {
    try {
      await this.checkIsOnInventoryPage();
      for (const selector of this.inventoryItems) {
        const item = this.page.locator(selector);
        if (await item.isVisible()) {
          await item.click();
        }
      }
    } catch (error) {
      throw new Error(`[addAllVisibleInventoryItems]: ${error}`);
    }
  }

  async addProductToCart(productLocator: keyof typeof InventoryPage.selectors): Promise<void> {
    try {
      const locator = this.page.locator(InventoryPage.selectors[productLocator]);
      await locator.waitFor({ state: 'visible', timeout: env.TIMEOUT });
      await locator.click();
    } catch (error) {
      throw new Error(`[addProductToCart]: ${error}`);
    }
  }

  async removeProductFromCart(productLocator: keyof typeof InventoryPage.selectors) {
    try {
      const locator = this.page.locator(InventoryPage.selectors[productLocator]);
      await locator.waitFor({ state: 'visible', timeout: env.TIMEOUT });
      await locator.click();
    } catch (error) {
      throw new Error(`[removeProductFromCart]: ${error}`);
    }
  }

  async sortItems(
    option: 'priceLowToHigh' | 'priceHighToLow' | 'nameAZ' | 'nameZA',
  ): Promise<void> {
    try {
      const mapping = {
        priceLowToHigh: 'lohi',
        priceHighToLow: 'hilo',
        nameAZ: 'az',
        nameZA: 'za',
      };
      await this.page.waitForURL(url.inventory);
      await this.sortDropdown.waitFor({
        state: 'visible',
        timeout: env.TIMEOUT,
      });
      await this.sortDropdown.selectOption(mapping[option]);
    } catch (error) {
      throw new Error(`[sortItems]: ${error}`);
    }
  }

  async getAllPrices(): Promise<number[]> {
    return await this.inventoryItemPrice.evaluateAll(els =>
      els.map(el => parseFloat(el.textContent.replace('$', '').trim())),
    );
  }

  async getAllNames(): Promise<string[]> {
    return await this.inventoryItemName.evaluateAll(els =>
      els.map(el => el.textContent.trim().toLowerCase()),
    );
  }

  async goToCart(): Promise<void> {
    await this.shoppingCartLink.click();
    await this.page.waitForURL(/cart/);
  }
}
